src/Agent/NewRelic/Agent/Extensions/NewRelic.Agent.Extensions/AwsSdk/SqsHelper.cs

```diff
@@ -48,25 +54,57 @@ public SqsAttributes(string url)
         public static ISegment GenerateSegment(ITransaction transaction, MethodCall methodCall, string url, MessageBrokerAction action)
         {
             var attr = new SqsAttributes(url);
-            return transaction.StartMessageBrokerSegment(methodCall, MessageBrokerDestinationType.Queue, action, VendorName, attr.QueueName);
+            var segment = transaction.StartMessageBrokerSegment(methodCall, MessageBrokerDestinationType.Queue, action, VendorName, attr.QueueName);
+            segment.GetExperimentalApi().MakeLeaf();
+
+            return segment;
         }
 
-        public static void InsertDistributedTraceHeaders(ITransaction transaction, dynamic webRequest)
+        public static void InsertDistributedTraceHeaders(ITransaction transaction, object sendMessageRequest)
         {
-            var setHeaders = new Action<dynamic, string, string>((wr, key, value) =>
+            var headersInserted = 0;
+
+            var setHeaders = new Action<object, string, string>((smr, key, value) =>
             {
-                var headers = wr.Headers as IDictionary<string, object>;
+                var getMessageAttributes = _getMessageAttributes ??=
+                    VisibilityBypasser.Instance.GeneratePropertyAccessor<IDictionary>(
+                        smr.GetType(), "MessageAttributes");
 
-                if (headers == null)
-                {
-                    headers = new Dictionary<string, object>();
-                    wr.Headers = headers;
-                }
+                var messageAttributes = getMessageAttributes(smr);
+
+                // SQS is limited to no more than 10 attributes; if we can't add up to 3 attributes, don't add any
+                if ((messageAttributes.Count + 3 - headersInserted) > 10)
```

Comment:
Please make sure there's a test case for this.

Reply:
Yep, working on tests for `SqsHelper` now, will add them to this PR.

Reply:
We should also confirm that CAT will also add no more than 3 attributes. It may also be possible to add a method that returns how many headers the current configuration would return.

Reply:
@nrcventura Do we really need to worry about CAT since it's supposed to be deprecated soon?